### PR TITLE
[Feature Request] Implement query placeholder

### PR DESCRIPTION
**Pull request by joker1007**
Because I want parameterized query execution, I implemented query placeholder feature.

## Placeholder syntax
Inspired by `ruby-pg` placeholder.
Placeholder syntax is `$<name>::<type>`

### Support types
- string (text)
- int
- date
- datetime
- float

### Query sample

```sql
SELECT * FROM users WHERE created_at > $created_at::date
```

## On Execution
If `Adhoq::Query#query` includes placeholder, execution form shows extra input field.

## Known issue
- use `sanitize_sql` method. `sanitize_sql` is AR class level protect method.
- type casting and form field is very heuristic.

### CODE REVIEW

File: app/models/adhoq/query.rb [hunk 1]
```diff
@@ -4,11 +4,61 @@ class Query < ActiveRecord::Base
...
+          end
+        end
+      end
+    end
+
+    def sanitized_query(query_parameters)
+      return query if parameters.empty?
+
+      casted = query_parameters.each_with_object(HashWithIndifferentAccess.new) do |(name, value_and_type), hash|
+        hash[name] = cast_query_parameter(value_and_type)
```
Comment (moro) [inline app/models/adhoq/query.rb]: More readable isn't it?
```
cast_query_parameter(*value_and_type.slice(:value, :type))
```

```
# l47..48 would go away
def cast_query_parameter(value, type)
```

**Comment by joker1007** [comment]
Because `Hash#slice` returns `Hash`,  cannot use `*`.
If this gem does not support ruby-1.9, I want to use keyword argument.
Or, How about that following code,

```ruby
      casted = query_parameters.each_with_object(HashWithIndifferentAccess.new) do |(name, value_and_type), hash|
        value = value_and_type[:value]
        type = value_and_type[:type]
        hash[name] = cast_query_parameter(value, type)
      end

```

http://api.rubyonrails.org/classes/Hash.html#method-i-slice

**Comment by moro** [comment]
Oh, sorry.

`Hash#values_at` is what I wanted to say.

**Comment by joker1007** [comment]
Ahh, I see.
Thanks!!

**Comment by myoung34** [comment]
any updates to this? I'm a fan of this =)

**Comment by takkanm** [comment]
@myoung34 I talked about this pull request with @joker1007 in meetup. So, He will refine this pull request. 
Please wait. :bow:

**Comment by joker1007** [comment]
## Update feature

I simplify placeholder spec, because it need not to take care of SQL injection.

## Placeholder syntax
Placeholder syntax is `$<name>` or `${<name>}` and
replaced input value as plain text

### Support types
discard type casting

### Query sample

```sql
SELECT * FROM users WHERE created_at > '$created_at'
```

need quotation in query string.

File: Gemfile.lock [hunk 2]
```diff
@@ -17,37 +17,37 @@ GEM
   remote: https://rubygems.org/
```
Comment (takkanm) [inline Gemfile.lock]: I think more better bundle updates is another pr. 
How about you?

**Comment by takkanm** [comment]
:ok_woman: thanks new feature.

**Pull request merged**